### ISSUE TITLE
[Bug]: The behavior of kmeans in SPU does not match kmeans in sklearns

### ISSUE THREAD

**Issue by winnylyc** [secretflow/spu#536]
### Issue Type

Currentness/Accuracy

### Modules Involved

Others

### Have you reproduced the bug with SPU HEAD?

Yes

### Have you searched existing issues?

Yes

### SPU Version

spu 0.7.0b0

### OS Platform and Distribution

Linux Ubuntu 22.04

### Python Version

3.10

### Current Behavior?

Hello, sorry for disturbing you.

When I use Kmeans implemented in sml/cluster, I encounter the problem that the behavior of it does not match kmeans implemented in sklearns with some inputs. I will include the code below.

### Standalone code to reproduce the issue

```Python
sim = spsim.Simulator.simple(
    3, spu_pb2.ProtocolKind.ABY3, spu_pb2.FieldType.FM64
)
def proc(x):
    model = KMEANS(n_clusters=3, n_samples=x.shape[0], max_iter=10)
    model.fit(x)
    return model._centers
X = jnp.array([[-4, -3, -2, -1]]).T
result = spsim.sim_jax(sim, proc)(X)
print("result\n", result)

# Compare with sklearn
from sklearn.cluster import KMeans

model = KMeans(n_clusters=3)
model.fit(X)
print("sklearn:\n", model.cluster_centers_)
```

### Relevant log output

```Shell
result
 [[-1.4999962]
 [ 0.       ]
 [-3.4999924]]
sklearn:
 [[-1. ]
 [-3.5]
 [-2. ]]
```

**Comment by deadlywing** [comment]
Hello，，thanks for reporting this.

The reason behind this is that sklearn runs the clustering algorithm several times（with different init centers） and use the best one.

However，sml just runs the procedure once for efficiency ，so it may get different outputs.

BTW，It's pretty straight to support this functionality，would you mind doing this job？

**Comment by winnylyc** [comment]
Thanks for your response!
I will try solving this problem.

**Comment by winnylyc** [comment]
I observed that another factor behind this difference is that default method to generate initial centers is different. I will also try adding a new initialization method.

**Comment by deadlywing** [comment]
> I observed that another factor behind this difference is that default method to generate initial centers is different. I will also try adding a new initialization method.

Exactly, sklearn use kmeans++ to decide init centroids, if you want to add this method, you MUST generate some random values before running in SPU.

**Comment by winnylyc** [comment]
Sorry, I don't catch the point. You mean I cannot use functions from jax.random in my implementation of kmeans++?

**Comment by deadlywing** [comment]
> Sorry, I don't catch the point. You mean I cannot use functions from jax.random in my implementation of kmeans++?

you can't generate random values in SPU runtime, you can refer to #80 . 
But you can use  jax.random.xxx in __init__, just like https://github.com/secretflow/spu/blob/0988eec76721b5419693b979e5709b7b4c3becde/sml/cluster/kmeans.py#L39-L41

**Comment by winnylyc** [comment]
I think I got your point. You mean that I need to generate all the random values I need in init function, and using jax.random.xxx in all the other functions will cause unexpected behavior?

**Comment by deadlywing** [comment]
> I think I got your point. You mean that I need to generate all the random values I need in init function, and using jax.random.xxx in all the other functions will cause unexpected behavior?

there are two cases in SML:
1. Generate these in `__init__` is OK only If the random values can be public (all paries can see these in plaintext)
2. If the random values are sensitive,  then you can feed these values in `fit` as input parameters, then other parties can't see these values.

**Comment by winnylyc** [comment]
So values of all the attributes in init function are public, right?
In addition, I think all the random values in kmeans++ can be public, so I will choose to generate them in init function. Is that appropriate?

**Comment by deadlywing** [comment]
> So values of all the attributes in init function are public, right?

Yes.

> In addition, I think all the random values in kmeans++ can be public, so I will choose to generate them in init function. Is that appropriate?

I think so.

**Comment by winnylyc** [comment]
Thanks for your reminder! It really helps a lot.

**Comment by winnylyc** [comment]
When I implemented the selection of the best centers, I encountered a behavior looks like a bug. 
Here is the code.
```
sim = spsim.Simulator.simple(
    3, spu_pb2.ProtocolKind.ABY3, spu_pb2.FieldType.FM64
)
def proc(x, centers):
    C = x.reshape((1, 1, x.shape[0], x.shape[1])) - centers.reshape(
        (centers.shape[0], centers.shape[1], 1, centers.shape[2])
    ) # shape [n_init, n_clusters, n_samples, n_features]
    distance =  jnp.sum(jnp.square(C), axis=3) # shape [n_init, n_clusters, n_samples]
    index = jnp.argmin(distance, axis=1) # shape [n_init, n_samples]
    inertia = jnp.sum(jnp.take(distance, index), axis=1) # shape [n_init]
    centers_best = centers[jnp.argmin(inertia)] # selection here, should choose centers with shape [n_clusters, n_features] from [n_init, n_clusters, n_features]
    return centers_best

X = jnp.ones((4, 2)) # shape [n_samples, n_features]
centers = jnp.ones((2, 3, 2)) # shape [n_init, n_clusters, n_features]

result = spsim.sim_jax(sim, proc)(X, centers)
print("result\n", result)
```
Here is the sample with n_init = 2, and it raises the following runtime error.
```
Traceback (most recent call last):
  File "smalltest.py", line 23, in <module>
    result = spsim.sim_jax(sim, proc)(X, centers)
  File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 169, in wrapper
    out_flat = sim(executable, *args_flat)
  File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 117, in __call__
    parties = [job.join() for job in jobs]
  File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 117, in <listcomp>
    parties = [job.join() for job in jobs]
  File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 47, in join
    raise self.exc
  File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 40, in run
    self.ret = self._target(*self._args, **self._kwargs)
  File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 106, in wrapper
    rt.run(executable)
  File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/api.py", line 44, in run
    return self._vm.Run(executable.SerializeToString())
RuntimeError: what: 
        [Enforce fail at libspu/core/ndarray_ref.cc:300] (end_indices[idx] <= shape()[idx]). Slice end at axis 0 = 3 is larger than input shape 2
Stacktrace:
#0 spu::kernel::hal::slice()+0x7fff4b00d93e
#1 spu::kernel::hlo::SecretDynamicSlice()+0x7fff4af236f2
#2 spu::kernel::hlo::DynamicSlice()+0x7fff4af24540
#3 spu::device::pphlo::dispatchOp<>()+0x7fff4a91d3fa
#4 spu::device::pphlo::dispatchOp<>()+0x7fff4a91f0e4
#5 spu::device::pphlo::dispatchOp<>()+0x7fff4a920f93
#6 spu::device::pphlo::dispatchOp<>()+0x7fff4a922fe9
#7 spu::device::pphlo::dispatchOp<>()+0x7fff4a9252f0
#8 spu::device::runBlock()+0x7fff4aa60f05
#9 spu::device::runRegion()+0x7fff4aa62f93
#10 spu::device::executeImpl()+0x7fff4a4497f0
#11 spu::RuntimeWrapper::Run()+0x7fff49c2e10c
#12 pybind11::cpp_function::initialize<>()::{lambda()#3}::_FUN()+0x7fff49c025b5
#13 pybind11::cpp_function::dispatcher()+0x7fff49bf785e
#14 cfunction_call_varargs+0x55555569000e

stacktrace: 
#0 spu::kernel::hal::slice()+0x7fff4b00d93e
#1 spu::kernel::hlo::SecretDynamicSlice()+0x7fff4af236f2
#2 spu::kernel::hlo::DynamicSlice()+0x7fff4af24540
#3 spu::device::pphlo::dispatchOp<>()+0x7fff4a91d3fa
#4 spu::device::pphlo::dispatchOp<>()+0x7fff4a91f0e4
#5 spu::device::pphlo::dispatchOp<>()+0x7fff4a920f93
#6 spu::device::pphlo::dispatchOp<>()+0x7fff4a922fe9
#7 spu::device::pphlo::dispatchOp<>()+0x7fff4a9252f0
#8 spu::device::runBlock()+0x7fff4aa60f05
#9 spu::device::runRegion()+0x7fff4aa62f93
#10 spu::device::executeImpl()+0x7fff4a4497f0
#11 spu::RuntimeWrapper::Run()+0x7fff49c2e10c
#12 pybind11::cpp_function::initialize<>()::{lambda()#3}::_FUN()+0x7fff49c025b5
#13 pybind11::cpp_function::dispatcher()+0x7fff49bf785e
#14 cfunction_call_varargs+0x55555569000e
```
The strange thing is that when n_init is set to any other positive number, it can return the expected result. Is this a bug or an expected behavior?

**Comment by tpppppub** [comment]
> When I implemented the selection of the best centers, I encountered a behavior looks like a bug. Here is the code.
> 
> ```
> sim = spsim.Simulator.simple(
>     3, spu_pb2.ProtocolKind.ABY3, spu_pb2.FieldType.FM64
> )
> def proc(x, centers):
>     C = x.reshape((1, 1, x.shape[0], x.shape[1])) - centers.reshape(
>         (centers.shape[0], centers.shape[1], 1, centers.shape[2])
>     ) # shape [n_init, n_clusters, n_samples, n_features]
>     distance =  jnp.sum(jnp.square(C), axis=3) # shape [n_init, n_clusters, n_samples]
>     index = jnp.argmin(distance, axis=1) # shape [n_init, n_samples]
>     inertia = jnp.sum(jnp.take(distance, index), axis=1) # shape [n_init]
>     centers_best = centers[jnp.argmin(inertia)] # selection here, should choose centers with shape [n_clusters, n_features] from [n_init, n_clusters, n_features]
>     return centers_best
> 
> X = jnp.ones((4, 2)) # shape [n_samples, n_features]
> centers = jnp.ones((2, 3, 2)) # shape [n_init, n_clusters, n_features]
> 
> result = spsim.sim_jax(sim, proc)(X, centers)
> print("result\n", result)
> ```
> 
> Here is the sample with n_init = 2, and it raises the following runtime error.
> 
> ```
> Traceback (most recent call last):
>   File "smalltest.py", line 23, in <module>
>     result = spsim.sim_jax(sim, proc)(X, centers)
>   File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 169, in wrapper
>     out_flat = sim(executable, *args_flat)
>   File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 117, in __call__
>     parties = [job.join() for job in jobs]
>   File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 117, in <listcomp>
>     parties = [job.join() for job in jobs]
>   File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 47, in join
>     raise self.exc
>   File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 40, in run
>     self.ret = self._target(*self._args, **self._kwargs)
>   File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/utils/simulation.py", line 106, in wrapper
>     rt.run(executable)
>   File "/home/ylipf/anaconda3/envs/spu/lib/python3.8/site-packages/spu/api.py", line 44, in run
>     return self._vm.Run(executable.SerializeToString())
> RuntimeError: what: 
>         [Enforce fail at libspu/core/ndarray_ref.cc:300] (end_indices[idx] <= shape()[idx]). Slice end at axis 0 = 3 is larger than input shape 2
> Stacktrace:
> #0 spu::kernel::hal::slice()+0x7fff4b00d93e
> #1 spu::kernel::hlo::SecretDynamicSlice()+0x7fff4af236f2
> #2 spu::kernel::hlo::DynamicSlice()+0x7fff4af24540
> #3 spu::device::pphlo::dispatchOp<>()+0x7fff4a91d3fa
> #4 spu::device::pphlo::dispatchOp<>()+0x7fff4a91f0e4
> #5 spu::device::pphlo::dispatchOp<>()+0x7fff4a920f93
> #6 spu::device::pphlo::dispatchOp<>()+0x7fff4a922fe9
> #7 spu::device::pphlo::dispatchOp<>()+0x7fff4a9252f0
> #8 spu::device::runBlock()+0x7fff4aa60f05
> #9 spu::device::runRegion()+0x7fff4aa62f93
> #10 spu::device::executeImpl()+0x7fff4a4497f0
> #11 spu::RuntimeWrapper::Run()+0x7fff49c2e10c
> #12 pybind11::cpp_function::initialize<>()::{lambda()#3}::_FUN()+0x7fff49c025b5
> #13 pybind11::cpp_function::dispatcher()+0x7fff49bf785e
> #14 cfunction_call_varargs+0x55555569000e
> 
> stacktrace: 
> #0 spu::kernel::hal::slice()+0x7fff4b00d93e
> #1 spu::kernel::hlo::SecretDynamicSlice()+0x7fff4af236f2
> #2 spu::kernel::hlo::DynamicSlice()+0x7fff4af24540
> #3 spu::device::pphlo::dispatchOp<>()+0x7fff4a91d3fa
> #4 spu::device::pphlo::dispatchOp<>()+0x7fff4a91f0e4
> #5 spu::device::pphlo::dispatchOp<>()+0x7fff4a920f93
> #6 spu::device::pphlo::dispatchOp<>()+0x7fff4a922fe9
> #7 spu::device::pphlo::dispatchOp<>()+0x7fff4a9252f0
> #8 spu::device::runBlock()+0x7fff4aa60f05
> #9 spu::device::runRegion()+0x7fff4aa62f93
> #10 spu::device::executeImpl()+0x7fff4a4497f0
> #11 spu::RuntimeWrapper::Run()+0x7fff49c2e10c
> #12 pybind11::cpp_function::initialize<>()::{lambda()#3}::_FUN()+0x7fff49c025b5
> #13 pybind11::cpp_function::dispatcher()+0x7fff49bf785e
> #14 cfunction_call_varargs+0x55555569000e
> ```
> 
> The strange thing is that only when first dimension (n_init) of centers is 2 will there be a runtime error. All the other settings of n_init will execute as expected. Is this a bug or an expected behavior?

Please try with the latest main branch code. This bug is fixed by the PR #532 .

**Comment by winnylyc** [comment]
Sorry for not testing with latest commit and disturbing you. The code can execute as expected in latest commit.

**Comment by winnylyc** [comment]
Solved with #546.
